src/replication.c

```diff
@@ -2308,6 +2476,443 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        server.repl_provisional_primary.dbid = dbid;
+
+        /* Now that we have the snapshot end-offset, we can ask for psync from that offset. Prepare the  
+         * main connection accordingly.*/
+        server.repl_transfer_s->state = CONN_STATE_CONNECTED;
+        server.repl_state = REPL_STATE_SEND_HANDSHAKE;
+        serverAssert(connSetReadHandler(server.repl_transfer_s, setupMainConnForPsync) != C_ERR);
+        setupMainConnForPsync(server.repl_transfer_s);
+
+        /* As the next block we will receive using this connection is the rdb, we need to prepare
+         * the connection accordingly */
+        serverAssert(connSetReadHandler(server.repl_rdb_transfer_s, readSyncBulkPayload) != C_ERR);
+        server.repl_transfer_size = -1;
+        server.repl_transfer_read = 0;
+        server.repl_transfer_last_fsync_off = 0;
+        server.repl_transfer_lastio = server.unixtime;
+
+        server.repl_rdb_conn_state = REPL_RDB_CONN_RDB_LOAD;
+        return;
+    }
+
+    no_response_error: 
+        /* Handle receiveSynchronousResponse() error when primary has no reply */
+        serverLog(LL_WARNING, "Primary did not respond to command during SYNC handshake");
+        /* Fall through to regular error handling */
+
+    error:
+        connClose(conn);
+        server.repl_transfer_s = NULL;
+        if (server.repl_rdb_transfer_s) {
+            connClose(server.repl_rdb_transfer_s);
+            server.repl_rdb_transfer_s = NULL;
+        }
+        if (server.repl_transfer_fd != -1)
+            close(server.repl_transfer_fd);
+        server.repl_transfer_fd = -1;
+        server.repl_state = REPL_STATE_CONNECT;
+        abortRdbConnectionSync();
+        return;
+
+    write_error: /* Handle sendCommand() errors. */
+        serverLog(LL_WARNING, "Sending command to primary in rdb connection replication handshake: %s", err);
+        sdsfree(err);
+        goto error;
+}
+
+/* Replication: Replica side.
+ * Initialize server.pending_repl_data infrastructure, we will allocate the buffer
+ * itself once we need it */
+void replDataBufInit(void) {
+    serverAssert(server.pending_repl_data.blocks == NULL);
+    server.pending_repl_data.len = 0;
+    server.pending_repl_data.peak = 0;
+    server.pending_repl_data.blocks = listCreate();
+    server.pending_repl_data.blocks->free = zfree;
+}
+
+/* Replication: Replica side.
+ * Track the local repl-data buffer streaming progress and serve clients from time to time */
+void replStreamProgressCallback(size_t offset, int readlen, time_t *last_progress_callback) {
+    time_t now = mstime();
+    if (server.loading_process_events_interval_bytes &&
+        (offset + readlen) / server.loading_process_events_interval_bytes > offset / server.loading_process_events_interval_bytes &&
+        now - *last_progress_callback > server.loading_process_events_interval_ms)
+    {
+        replicationSendNewlineToPrimary();
+        processEventsWhileBlocked();
```

Comment:
I don't think we track progress and handle clients in `rdbLoad`. Can you talk about the rationale?

Reply:
We track the progress of RDB loading using https://github.com/valkey-io/valkey/blob/7415a576a8c49fdb946ac45aa66f4a462872d277/src/rdb.c#L2923

The replica invokes this method from within `rioRead` as it reads the RDB file.
In particular, we utilize this in `replStreamProgressCallback`. This is crucial when the replica has accumulated a substantial replication change buffer. By tracking progress, we prevent other instances from mistakenly assuming the replica is unresponsive or has failed.

Reply:
Ah - thanks! for some reason, I've always assumed it is for check-summing only.